#182210045 user role settings and assigning

nyakamwe opened the pull request:
**What does this PR do?**
It helps the super administrator to assign roles to any staff member.

**Description of Task to be completed?**
    Create a Role model
    Seeds all application user roles
    Associates(relates) each user to a corresponding role
    Make update user role functionality by allowing only logged-in super admin to perform that.

**How should this be manually tested?**
After cloning this repo:
1. Create a `.env` file referring to the `.env.example` file in the root folder and provide values for those keys.
2. Run `npm install` for installing all dependencies.
3. `npm run dbseed` to seed application's user roles

**What are the relevant pivotal tracker stories?**
[ #182210045](https://www.pivotaltracker.com/n/projects/2569261/stories/182210045)

Himbaza123h committed:
change-scripts

rebase eslint config file

munezerobagira committed:
ft: logout functionality

- add the user session model

munezerobagira committed:
ft: logout functionality

- add the user session model

Himbaza123h committed:
feat(Authentication): Generating JWT for user

rebase in User model

rebase

resolving conflicts

Himbaza123h committed:
feat(Authentication): Generating JWT for user

rebase in User model

rebase

resolving conflicts

Himbaza123h committed:
feat(Authentication): Generating JWT for user

rebase in User model

Himbaza123h committed:
feat(Authentication): correction based on PR's comments

Himbaza123h committed:
feat(Authentication): resolving conflicts

rebasing

Himbaza123h committed:
feat(Authentication): resolving conflicts

Himbaza123h committed:
feat(Authentication): resolving conflicts

resolving conficts

Himbaza123h committed:
feat(Auth check middlewares): Coveralls test

resolving conflicts

Himbaza123h committed:
feat(Auth check middlewares): Coveralls test

resolving conflicts

Himbaza123h committed:
feat(Auth check middlewares): Coveralls test

resolving conflicts

Himbaza123h committed:
feat(Auth check middlewares): Coveralls test

resolving conflicts

rebase

munezerobagira committed:
feat(logout) implement logout functionality

- updated session model

added the test

munezerobagira committed:
feat(logout) implement logout functionality

- updated session model

added the test

nyakamwe committed:
feat(user-roles-settings): implement user role assigning

- Admin has priviliges of updating user role
- Seed Admin at very beginning of application
- documentation of assignRole endpoint
- Tests of updating user role
- Check if user is admin or not
- Seed roles of application's user

[User Role Settings #182210045]

rebasing

rebasing

nyakamwe committed:
feat(user-roles-settings): implement user role assigning

- Admin has priviliges of updating user role
- Seed Admin at very beginning of application
- documentation of assignRole endpoint
- Tests of updating user role
- Check if user is admin or not
- Seed roles of application's user

[User Role Settings #182210045]

nyakamwe committed:
feat(user-roles-settings): implement user role assigning

- Admin has priviliges of updating user role
- Seed Admin at very beginning of application
- documentation of assignRole endpoint
- Tests of updating user role
- Check if user is admin or not
- Seed roles of application's user

[User Role Settings #182210045]

rebase routes

resolving conficts

nyakamwe committed:
feat(user-roles-settings): implement user role assigning

- Admin has priviliges of updating user role
- Seed Admin at very beginning of application
- documentation of assignRole endpoint
- Tests of updating user role
- Check if user is admin or not
- Seed roles of application's user

[User Role Settings #182210045]

rebasing

rebasing

habaumugisha1 commented on package.json:
```diff
@@ -4,15 +4,17 @@
+    "test": "cross-env NODE_ENV=test yarn run unmigrate && cross-env NODE_ENV=test yarn run migrate && cross-env NODE_ENV=test && cross-env NODE_ENV=test yarn run dbseed && cross-env NODE_ENV=test nyc --reporter=text --reporter=html  mocha --recursive --require @babel/polyfill --require @babel/register ./src/test/*.js --timeout 30000 --exit",
     "coverage": "nyc  --reporter=text-lcov --reporter=lcov npm run test | coveralls",
-    "start": "cross-env NODE_ENV=production npm run dbmigrate && node lib/index.js",
+    "start": "cross-env NODE_ENV=production npm run dbmigrate && cross-env NODE_ENV=production npm run dbseed  && node lib/index.js",
```
we don't need this 
## npm run dbseed script

habaumugisha1 commented on src/database/migrations/20220601181025-create-user.js:
```diff
@@ -30,8 +30,8 @@ module.exports = {
-        type: Sequelize.ENUM("admin", "requester", "manager", "traveler"),
-        defaultValue: "requester",
+        type: Sequelize.UUID,
+        defaultValue: "7adae2f1-4d35-470d-8512-1b9634330a9e",
```
use something like this
## default: Sequelize.fn('uuid_generate_v4'),

nyakamwe replied:
This is a defaut value of user which is requester role id @habaumugisha1

habaumugisha1 commented on src/middlewares/isSuperAdmin.js:
```diff
@@ -0,0 +1,28 @@
+		if(error) return res.status(403).json({message:"Invalid Token"})
+		// console.log(user);
+		
+        if(user.user_role !== "013dddd7-2769-4de6-8fc3-7aa527114879"){
```
why are you using this value here?

nyakamwe replied:
@habaumugisha1  This is in middleware that protect routes to be accessed by  only admin, so this value refer to id of admin user_role

nyakamwe committed:
feat(user-roles-settings): fallback to npm

- replace yarn npm
- create role in before hook during testing

[User Role Settings #182210045]

munezerobagira committed:
feat(logout) logout functionality

- add script to run undo migration after test

add test for logout

Himbaza123h committed:
feat(Auth check middleware):

resolving conflicts

rebasing

rebasing

Himbaza123h committed:
feat(Auth check middleware):

resolving conflicts

rebasing

rebasing

Himbaza123h committed:
feat(Auth check middleware):

resolving conflicts

Himbaza123h committed:
feat(Auth check middleware): fixing errors with circle ci checks

rebasing

nyakamwe committed:
feat(user-roles-settings): add functionality in isSuperAdmin middleware

munezerobagira committed:
feat(logout) logout functionality

- added the routes for sessions

nyakamwe committed:
feat(user-roles-settings): retouch from rebase

nyakamwe committed:
feat(user-roles-settings): retouch from rebase

rebasing

resolving

munezerobagira committed:
feat(logout) logout functionality

- added more test

nyakamwe committed:
feat(user-roles-settings): retaouch on failing test

nyakamwe committed:
feat(user-roles-settings): require dotenv in sequelize file config

nyakamwe committed:
feat(user-roles-settings): resolve conflicts

resolving conflict

nyakamwe committed:
feat(user-roles-settings): resolve conflicts

nyakamwe committed:
feat(user-roles-settins): fall to remote db

nyakamwe committed:
feat(user-roles-settins): fall to remote db

nyakamwe committed:
Merge branch 'develop' into feat-user-role-settings-#182210045

resolving conflicts

nyakamwe committed:
feat(user-roles-settings): add environment to dev script

habaumugisha1 commented on package.json:
```diff
@@ -4,17 +4,20 @@
     "coverage": "nyc  --reporter=text-lcov --reporter=lcov npm run test | coveralls",
-    "start": "cross-env NODE_ENV=production npm run dbmigrate && node lib/index.js",
-    "dev": "nodemon --exec npx babel-node src/index.js",
+    "start": "cross-env NODE_ENV=production npm run migrate && cross-env NODE_ENV=production npm run dbseed  && node lib/index.js",
```
You don't need to seed the data every time when app start it can cause app to crash because of the duplication emails on the user

we only need to run migration

habaumugisha1 commented on src/controllers/usersControllers.js:
```diff
@@ -0,0 +1,40 @@
+    try {
+        const user = await UserService.checkUser(email);
+        if(user == null){
+            res.status(404).json({status:404, message:"user doesn't exist"});
```
you can write in one line 

```if(user == null) return res.status(404).json({status:404, message:"user doesn't exist"}); ```

habaumugisha1 commented on src/controllers/usersControllers.js:
```diff
@@ -0,0 +1,40 @@
+        const user = await UserService.checkUser(email);
+        if(user == null){
+            res.status(404).json({status:404, message:"user doesn't exist"});
+            return false;
```
and then remove this return false after use the line above

munezerobagira committed:
feat(logout) logout functionality

- fix test

munezerobagira committed:
feat(logout) logout functionality

- fix test

munezerobagira committed:
feat(logout) logout functionality

- debug the test online

nyakamwe committed:
feat(user-roles-settings): refactor user controller

nyakamwe committed:
feat(user-roles-settings): work  on feedback

nyakamwe committed:
feat(user-roles-settings): work  on feedback

nyakamwe committed:
feat(user-roles-settings): rebasing with develop

resolving

nyakamwe committed:
feat(user-roles-settings): do retouch on tests

nyakamwe committed:
feat(user-roles-settings): resolve index of documenation

nyakamwe committed:
feat(Authentication): correction based on PR's comments

nyakamwe committed:
feat(Auth check middleware):

resolving conflicts

habaumugisha1 commented:
resolve these conflicts @nyakamwe

nyakamwe committed:
documentation(swagger): add Swagger docs for acc signup and verification

nyakamwe committed:
test(emailing): allows usage of invalid sendgrid key for tests

nyakamwe committed:
feat(user-roles-settings): implement user role assigning

- Admin has priviliges of updating user role
- Seed Admin at very beginning of application
- documentation of assignRole endpoint
- Tests of updating user role
- Check if user is admin or not
- Seed roles of application's user

[User Role Settings #182210045]

nyakamwe committed:
feat(user-roles-settings): implement user role assigning

- Admin has priviliges of updating user role
- Seed Admin at very beginning of application
- documentation of assignRole endpoint
- Tests of updating user role
- Check if user is admin or not
- Seed roles of application's user

[User Role Settings #182210045]

rebase routes

resolving conficts

nyakamwe committed:
feat(user-roles-settings): retouch from rebase

nyakamwe committed:
feat(user-roles-settings): retaouch on failing test

nyakamwe committed:
feat(user-roles-settings): retaouch on failing test

nyakamwe committed:
feat(user-roles-settings): resolve conflicts

nyakamwe committed:
feat(Authentication): Generating JWT for user

rebase in User model

rebase

resolving conflicts

nyakamwe committed:
feat(Authentication): Generating JWT for user

rebase in User model

rebase

resolving conflicts

nyakamwe committed:
feat(user-roles-settings): do retouch on tests

nyakamwe committed:
feat(user-roles-settings): increase timeout

nyakamwe committed:
Merge branch 'develop' into feat-user-role-settings-#182210045

istafrica-dennw commented on src/middlewares/isSuperAdmin.js:
```diff
@@ -0,0 +1,23 @@
+
+export function isSuperAdmin (req, res, next){
+
+    const authHeader = req.headers['authorization']
```
can we add this logic in auth middleware (I guess it there) and add the token in headers that were passing via next()

munezerobagira committed:
feat(logout) logout functionality

- fix filename

nyakamwe committed:
feat(user-roles-settings): refactor auth middleware

munezerobagira committed:
feat(logout) logout functionality

- change from x-auth-token to authorization

nyakamwe committed:
feat(user-roles-settings): implement user role assigning

- Admin has priviliges of updating user role
- Seed Admin at very beginning of application
- documentation of assignRole endpoint
- Tests of updating user role
- Check if user is admin or not
- Seed roles of application's user

[User Role Settings #182210045]

nyakamwe committed:
feat(user-roles-settings): fallback to npm

- replace yarn npm
- create role in before hook during testing

[User Role Settings #182210045]

nyakamwe committed:
feat(user-roles-settings): add functionality in isSuperAdmin middleware

nyakamwe committed:
feat(Authentication): Generating JWT for user

rebase in User model

nyakamwe committed:
feat(Auth check middleware):

resolving conflicts

nyakamwe committed:
feat(user-roles-settings): implement user role assigning

- Admin has priviliges of updating user role
- Seed Admin at very beginning of application
- documentation of assignRole endpoint
- Tests of updating user role
- Check if user is admin or not
- Seed roles of application's user

[User Role Settings #182210045]

rebase routes

resolving conficts

nyakamwe committed:
feat(user-roles-settings): retouch from rebase

nyakamwe committed:
feat(user-roles-settings): retaouch on failing test

nyakamwe committed:
feat(user-roles-settings): work  on feedback

nyakamwe committed:
feat(Auth check middleware):

resolving conflicts

rebasing

rebasing

nyakamwe committed:
feat(user-roles-settings): retouch from rebase

rebasing

resolving

nyakamwe committed:
feat(user-roles-settings): do retouch on tests

nyakamwe committed:
documentation(swagger): add Swagger docs for acc signup and verification

nyakamwe committed:
feat(user-roles-settings): implement user role assigning

- Admin has priviliges of updating user role
- Seed Admin at very beginning of application
- documentation of assignRole endpoint
- Tests of updating user role
- Check if user is admin or not
- Seed roles of application's user

[User Role Settings #182210045]

nyakamwe committed:
feat(Auth check middleware):

resolving conflicts

nyakamwe committed:
feat(user-roles-settings): resolve conflicts

nyakamwe committed:
feat(Auth check middleware):

resolving conflicts

rebasing

rebasing

nyakamwe committed:
feat(user-roles-settings): refactor auth middleware

nyakamwe committed:
feat(user-roles-settings): retouch on test and resolve conflicts

nyakamwe committed:
feat(user-roles-settings): wrap up the user role settings"

nyakamwe committed:
feat(user-roles-settings): alter start script
